python/mxnet/ndarray/sparse_ndarray.py

```diff
@@ -0,0 +1,637 @@
+    >>> a.data.asnumpy()
+    array([ 1.,  2.,  3.], dtype=float32)
+    """
+
+    def __reduce__(self):
+        return CSRNDArray, (None,), super(CSRNDArray, self).__getstate__()
+
+    @property
+    def indices(self):
+        """The indices array of the SparseNDArray with `csr` storage type.
+        This generates a deep copy of the column indices of the current `csr` matrix.
+
+        Returns
+        -------
+        NDArray
+            This SparseNDArray's indices array.
+        """
+        return self._aux_data(1)
+
+    @property
+    def indptr(self):
+        """The indptr array of the SparseNDArray with `csr` storage type.
+        This generates a deep copy of the `indptr` of the current `csr` matrix.
+
+        Returns
+        -------
+        NDArray
+            This SparseNDArray's indptr array.
+        """
+        return self._aux_data(0)
+
+# pylint: disable=abstract-method
+class RowSparseNDArray(SparseNDArray):
+    """A RowSparseNDArray is typically used to represent a subset of a larger
+    NDArray  with `default` of shape [LARGE0, D1, .. , DN] where LARGE0 >> D0. The values
+    in indices are the indices in the first dimension of the slices that have been extracted from
+    the larger NDArray. The indices are expected to be sorted in ascending order.
+
+    The corresponding NDArray ``dense`` with `default` storage represented by a ``rsp``
+    RowSparseNDArray
+
+    ``dense[rsp.indices[i], :, :, :, ...] = rsp.values[i, :, :, :, ...]``
+
+    RowSparseNDArray is used principally in the definition of gradients for operations
+    that have sparse gradients (e.g. dot with sparse inputs).
+
+    Examples
+    --------
+    >>> import mxnet as mx
+    >>> dense = mx.nd.array([[1,2],[0,0],[3,0],[0,0]])
+    >>> rsp = dense._to_rsp()
+    >>> rsp.indices.asnumpy()
+    array([0, 2], dtype=int32)
+    >>> rsp.data.asnumpy()
+    array([[ 1.,  2.],
+           [ 3.,  0.]], dtype=float32)
+    """
+    def __reduce__(self):
+        return RowSparseNDArray, (None,), super(RowSparseNDArray, self).__getstate__()
+
+    @property
+    def indices(self):
+        """The indices array of the SparseNDArray with `row_sparse` storage type.
+        This generates a deep copy of the row indices of the current row-sparse matrix.
+
+        Returns
+        -------
+        NDArray
+            This SparseNDArray's indices array.
+        """
+        return self._aux_data(0)
+
+
+def _prepare_src_array(src, dtype, default_dtype):
+    if isinstance(src, NDArray):
+        dtype = src.dtype if dtype is None else dtype
+    else:
+        dtype = default_dtype if dtype is None else dtype
+        if not isinstance(src, np.ndarray):
+            try:
+                src = np.array(src, dtype=dtype)
+            except:
+                raise TypeError('values must be array like object')
+    return src, dtype
+
+
+def csr(data, indptr, indices, shape, ctx=None, dtype=None, indptr_type=None, indices_type=None):
```

Comment:
actually if we want to make it longer I prefer csr_matrix instead of csr_array since it's always going to be 2D and that's how scipy call it